File: linux/DotsHarness/src/PluginRuntime/InMemoryRegistries.cs
// Copyright (c) 2026 DOTS
// Plugin composition model derived from DeepSeek Harness.
// Copyright (c) 2026 DeepSeek. MIT. See NOTICE.

using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using HarnessPluginKit;

namespace PluginRuntime;

public static class FiberLocal
{
    private static readonly AsyncLocal<string> OwnerSlot = new();
    public static string Owner
    {
        get => string.IsNullOrEmpty(OwnerSlot.Value) ? "host" : OwnerSlot.Value;
        set => OwnerSlot.Value = value;
    }

    public static T WithOwner<T>(string owner, Func<T> body)
    {
        var previous = OwnerSlot.Value;
        OwnerSlot.Value = owner;
        try { return body(); }
        finally { OwnerSlot.Value = previous; }
    }

    public static void WithOwner(string owner, Action body)
    {
        var previous = OwnerSlot.Value;
        OwnerSlot.Value = owner;
        try { body(); }
        finally { OwnerSlot.Value = previous; }
    }
}

public sealed class InMemoryPromptRegistry : IPromptRegistry
{
    private readonly List<PromptSection> _items = new();

    public void RetractOwner(string owner) => _items.RemoveAll(i => i.Owner == owner);

    public Action Section(string name, int order, string text)
    {
        var owner = FiberLocal.Owner;
        _items.RemoveAll(i => i.Name == name && i.Owner == owner);
        _items.Add(new PromptSection(name, order, text, owner));
        return () => _items.RemoveAll(i => i.Name == name && i.Owner == owner);
    }

    public string AssembledText() =>
        string.Join("\n\n", _items.OrderBy(i => i.Order).Select(i => i.Text));

    public IReadOnlyList<PromptSection> Sections() =>
        _items.OrderBy(i => i.Order).ToList();
}

public sealed class InMemoryToolRegistry : IToolRegistry
{
    private sealed record Stored(RegisteredTool Meta, Func<IReadOnlyDictionary<string, string>, Task<string>> Execute);

    private readonly Dictionary<string, Stored> _items = new(StringComparer.Ordinal);

    public void RetractOwner(string owner)
    {
        foreach (var key in _items.Where(kv => kv.Value.Meta.Owner == owner).Select(kv => kv.Key).ToList())
        {
            _items.Remove(key);
        }
    }

    public Action Register(
        string name,
        string description,
        IReadOnlyList<ToolParameter> parameters,
        Func<IReadOnlyDictionary<string, string>, Task<string>> execute)
    {
        var owner = FiberLocal.Owner;
        _items[name] = new Stored(new RegisteredTool(name, description, parameters, owner), execute);
        return () =>
        {
            if (_items.TryGetValue(name, out var stored) && stored.Meta.Owner == owner)
            {
                _items.Remove(name);
            }
        };
    }

    public IReadOnlyList<RegisteredTool> Tools() =>
        _items.Values.Select(v => v.Meta).OrderBy(t => t.Name).ToList();

    public Task<string> CallAsync(string name, IReadOnlyDictionary<string, string> arguments)
    {
        if (!_items.TryGetValue(name, out var stored))
        {
            throw PluginException.ApplyFailed($"unknown tool {name}");
        }
        return stored.Execute(arguments);
    }
}

public sealed class InMemorySlotRegistry : ISlotRegistry, INotifyPropertyChanged
{
    private readonly List<SlotRegistration> _items = new();

    public event PropertyChangedEventHandler? PropertyChanged;

    public void RetractOwner(string owner)
    {
        var removed = _items.RemoveAll(i => i.Owner == owner);
        if (removed > 0) OnChanged();
    }

    public Action Inject(string slot, string id, int order, string label, Func<object> viewFactory)
    {
        var owner = FiberLocal.Owner;
        _items.RemoveAll(i => i.Slot == slot && i.Id == id && i.Owner == owner);
        _items.Add(new SlotRegistration(id, slot, order, label, owner, viewFactory));
        OnChanged();
        return () =>
        {
            if (_items.RemoveAll(i => i.Slot == slot && i.Id == id && i.Owner == owner) > 0)
            {
                OnChanged();
            }
        };
    }

    public IReadOnlyList<SlotRegistration> Occupants(string slot) =>
        _items.Where(i => i.Slot == slot).OrderBy(i => i.Order).ToList();

    public IReadOnlyList<SlotRegistration> All() =>
        _items.OrderBy(i => i.Slot).ThenBy(i => i.Order).ThenBy(i => i.Id).ToList();

    private void OnChanged() => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(All)));
}

public sealed class InMemorySettingsRegistry : ISettingsRegistry
{
    private readonly Dictionary<string, JsonValue> _values;
    private readonly Action<IReadOnlyDictionary<string, JsonValue>>? _persist;

    public InMemorySettingsRegistry(
        IReadOnlyDictionary<string, JsonValue>? values = null,
        Action<IReadOnlyDictionary<string, JsonValue>>? persist = null)
    {
        _values = values is null
            ? new Dictionary<string, JsonValue>(StringComparer.Ordinal)
            : new Dictionary<string, JsonValue>(values, StringComparer.Ordinal);
        _persist = persist;
    }

    public JsonValue? Get(string key) => _values.TryGetValue(key, out var value) ? value : null;

    public void Set(string key, JsonValue value)
    {
        _values[key] = value;
        _persist?.Invoke(_values);
    }

    public IReadOnlyDictionary<string, JsonValue> Snapshot() =>
        new ReadOnlyDictionary<string, JsonValue>(new Dictionary<string, JsonValue>(_values, StringComparer.Ordinal));
}

public sealed class InMemoryEventBus : IEventBus
{
    private readonly Dictionary<string, Dictionary<Guid, Action<object?>>> _handlers = new(StringComparer.Ordinal);

    public Action On(string name, Action<object?> handler)
    {
        var token = Guid.NewGuid();
        if (!_handlers.TryGetValue(name, out var bucket))
        {
            bucket = new Dictionary<Guid, Action<object?>>();
            _handlers[name] = bucket;
        }
        bucket[token] = handler;
        return () =>
        {
            if (_handlers.TryGetValue(name, out var existing))
            {
                existing.Remove(token);
            }
        };
    }

    public void Emit(string name, object? payload)
    {
        if (!_handlers.TryGetValue(name, out var bucket)) return;
        foreach (var handler in bucket.Values.ToList()) handler(payload);
    }
}

public abstract class ObservableObject : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;

    protected bool SetProperty<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        OnPropertyChanged(name);
        return true;
    }

    protected void OnPropertyChanged([CallerMemberName] string? name = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

File: macos/DotsHarness/Sources/PluginRuntime/InMemoryRegistries.swift
// Copyright (c) 2026 DOTS
// In-memory plugin registries for Dots Harness.

import Foundation
import SwiftUI
import HarnessPluginKit

@MainActor
public final class InMemoryPromptRegistry: PromptRegistry {
    private var items: [PromptSection] = []

    public init() {}

    public func retractOwner(_ owner: String) {
        items.removeAll { $0.owner == owner }
    }

    @discardableResult
    public func section(name: String, order: Int, text: String) -> () -> Void {
        let owner = FiberLocal.owner
        items.removeAll { $0.name == name && $0.owner == owner }
        items.append(PromptSection(name: name, order: order, text: text, owner: owner))
        return { [weak self] in
            self?.items.removeAll { $0.name == name && $0.owner == owner }
        }
    }

    public func assembledText() -> String {
        items.sorted { $0.order < $1.order }.map(\.text).joined(separator: "\n\n")
    }

    public func sections() -> [PromptSection] {
        items.sorted { $0.order < $1.order }
    }
}

@MainActor
public final class InMemoryToolRegistry: ToolRegistry {
    private struct Stored {
        var meta: RegisteredTool
        var execute: ([String: String]) async throws -> String
    }

    private var items: [String: Stored] = [:]

    public init() {}

    public func retractOwner(_ owner: String) {
        items = items.filter { $0.value.meta.owner != owner }
    }

    @discardableResult
    public func register(
        name: String,
        description: String,
        parameters: [ToolParameter],
        execute: @escaping ([String: String]) async throws -> String
    ) -> () -> Void {
        let owner = FiberLocal.owner
        items[name] = Stored(
            meta: RegisteredTool(name: name, description: description, parameters: parameters, owner: owner),
            execute: execute
        )
        return { [weak self] in
            if self?.items[name]?.meta.owner == owner {
                self?.items.removeValue(forKey: name)
            }
        }
    }

    public func tools() -> [RegisteredTool] {
        items.values.map(\.meta).sorted { $0.name < $1.name }
    }

    public func call(_ name: String, arguments: [String: String]) async throws -> String {
        guard let stored = items[name] else {
            throw PluginError.applyFailed("unknown tool \(name)")
        }
        return try await stored.execute(arguments)
    }
}

@MainActor
public final class InMemorySlotRegistry: ObservableObject, SlotRegistry {
    @Published private var items: [SlotRegistration] = []

    public init() {}

    public func retractOwner(_ owner: String) {
        items.removeAll { $0.owner == owner }
    }

    @discardableResult
    public func inject(
        _ slot: String,
        id: String,
        order: Int,
        label: String,
        @ViewBuilder view: () -> some View
    ) -> () -> Void {
        let owner = FiberLocal.owner
        if let existing = items.first(where: { $0.slot == slot && $0.id == id && $0.owner != owner }) {
            assertionFailure("slot collision \(slot)/\(existing.id)")
        }
        items.removeAll { $0.slot == slot && $0.id == id && $0.owner == owner }
        items.append(SlotRegistration(id: id, slot: slot, order: order, label: label, owner: owner, view: AnyView(view())))
        return { [weak self] in
            self?.items.removeAll { $0.slot == slot && $0.id == id && $0.owner == owner }
        }
    }

    public func occupants(in slot: String) -> [SlotRegistration] {
        items.filter { $0.slot == slot }.sorted { $0.order < $1.order }
    }

    public func all() -> [SlotRegistration] {
        items.sorted { lhs, rhs in
            (lhs.slot, lhs.order, lhs.id) < (rhs.slot, rhs.order, rhs.id)
        }
    }
}

@MainActor
public final class InMemorySettingsRegistry: SettingsRegistry {
    private var values: [String: JSONValue]
    private let persist: (([String: JSONValue]) -> Void)?

    public init(values: [String: JSONValue] = [:], persist: (([String: JSONValue]) -> Void)? = nil) {
        self.values = values
        self.persist = persist
    }

    public func get(_ key: String) -> JSONValue? { values[key] }

    public func set(_ key: String, _ value: JSONValue) {
        values[key] = value
        persist?(values)
    }

    public func snapshot() -> [String: JSONValue] { values }
}

@MainActor
public final class InMemoryEventBus: EventBus {
    private var handlers: [String: [UUID: (Any?) -> Void]] = [:]

    public init() {}

    @discardableResult
    public func on(_ name: String, _ handler: @escaping (Any?) -> Void) -> () -> Void {
        let token = UUID()
        handlers[name, default: [:]][token] = handler
        return { [weak self] in
            self?.handlers[name]?[token] = nil
        }
    }

    public func emit(_ name: String, _ payload: Any?) {
        handlers[name]?.values.forEach { $0(payload) }
    }
}

enum FiberLocal {
    @TaskLocal static var owner: String = "host"
}
